### PR TITLE
[Training] Adam Optimizer

### PR DESCRIPTION
**Pull request by wschin**
**PR #2314 is a single place for reviewing the whole training story.**

A common signature shared by [Pytroch](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html)'s and [TF](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer)'s Adam.

Design verification script (it shows that the proposed Adam signature covers both of TF and Pytorch):
```python
import itertools
import numpy as np
import torch
from torch import optim
from torch import nn
import torch.random as rnd

iteration_count = 3
n = 10 # number of features
l = 5 # number of data points (aka batch size)

X_ = torch.randn(l, n)
Y_ = torch.randn(l, 1)

def apply_adam(t, r, x, g, v, h, norm_coefficient, alpha, beta, epsilon):  # type: ignore
    # Add gradient of regularization term.
    g_regularized = norm_coefficient * x + g
    # Update momentum.
    v_new = alpha * v + (1 - alpha) * g_regularized
    # Update second-order momentum.
    h_new = beta * h + (1 - beta) * (g_regularized * g_regularized)
    # Compute element-wise square root.
    h_sqrt = np.sqrt(h_new) + epsilon
    # Adjust learning rate.
    t = t + 1
    r_adjusted = r * np.sqrt(1 - beta**t) / (1 - alpha**t)
    # Apply Adam update rule.
    x_new = x - r_adjusted * (v_new / h_sqrt)
    return x_new, v_new, h_new

def show_pytorch(lr, alpha, beta, weight_decay, epsilon):
    X = X_.clone()
    Y = Y_.clone()

    rnd.manual_seed(0)
    model = nn.Sequential(
        nn.Linear(n, 1, bias=False)
    )

    loss_fn = nn.MSELoss(reduction='sum')

    solver = optim.Adam(model.parameters(), lr=lr, betas=[alpha, beta],
            weight_decay=weight_decay, eps=epsilon)

    results = []
    for t in range(iteration_count):
        Y_pred = model(X)
        loss = loss_fn(Y_pred, Y)
        results.append(float(loss))
        model.zero_grad()
        loss.backward()
        solver.step()
    return results

def show_tensorflow(lr, alpha, beta, epsilon):
    import tensorflow as tf
    rnd.manual_seed(0)
    layer = nn.Linear(n, 1, bias=False)

    X = tf.placeholder('float', shape=[l, n])
    W = tf.Variable(torch.Tensor.numpy(layer.weight.detach()))
    Y = tf.placeholder('float', shape=[l, 1])
    Y_pred = tf.matmul(X, W, transpose_b=True)
    loss = tf.reduce_sum(tf.square(Y - Y_pred))
    optimizer = tf.train.AdamOptimizer(learning_rate=lr, beta1=alpha, beta2=beta, epsilon=epsilon)
    minimizer = optimizer.minimize(loss)

    sess = tf.Session()
    init = tf.global_variables_initializer()
    sess.run(init)

    results = []
    for t in range(iteration_count):
        loss_value = sess.run(loss, {X: X_, Y: Y_})
        results.append(float(loss_value ))
        sess.run([minimizer], {X: X_, Y: Y_})

    return results

def show_mine(lr, alpha, beta, weight_decay, epsilon):
    X = X_.clone()
    Y = Y_.clone()

    rnd.manual_seed(0)
    model = nn.Sequential(
        nn.Linear(n, 1, bias=False)
    )

    loss_fn = nn.MSELoss(reduction='sum')

    solver = optim.Adam(model.parameters(), lr=lr, betas=[alpha, beta],
            weight_decay=weight_decay, eps=epsilon)

    results = []
    for t in range(iteration_count):
        Y_pred = model(X)
        loss = loss_fn(Y_pred, Y)
        results.append(float(loss))
        model.zero_grad()
        loss.backward()

        with torch.no_grad():
            for param in model.parameters():
                if 'exp_avg' not in solver.state[param]:
                    solver.state[param]['exp_avg'] = torch.zeros_like(param)
                if 'exp_avg_sq' not in solver.state[param]:
                    solver.state[param]['exp_avg_sq'] = torch.zeros_like(param)

                new_tensor, new_v, new_h = apply_adam(t=t,
                        r=lr, x=param.data, g=param.grad.data,
                        v=solver.state[param]['exp_avg'].data,
                        h=solver.state[param]['exp_avg_sq'].data,
                        norm_coefficient=weight_decay,
                        alpha=alpha, beta=beta, epsilon=epsilon)

                solver.state[param]['exp_avg'].data = new_v.data
                solver.state[param]['exp_avg_sq'].data = new_h.data
                param.data = new_tensor.data

    return results


# Compare TF Adam and its ONNX counterpart.
lr_pool = [0.1, 0.001]
alpha_pool = [0.89, 0.8, 0.01]
beta_pool = [0.99]
weight_decay_pool = [0]
epsilon_pool = [1e-7, 1e-1]
for lr_, alpha_, beta_, weight_decay_, epsilon_ in itertools.product(lr_pool, alpha_pool, beta_pool, weight_decay_pool, epsilon_pool):
    tf_result = show_tensorflow(lr=lr_, alpha=alpha_, beta=beta_, epsilon=epsilon_)
    mine_result = show_mine(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, epsilon=epsilon_)
    assert np.allclose(tf_result, mine_result)

# Compare Pytorch Adam and its ONNX counterpart.
lr_pool = [0.1, 0.001]
alpha_pool = [0.89, 0.8, 0.01]
beta_pool = [0.99, 0.2, 0.5]
weight_decay_pool = [0, 0.01, 0.1, 1]
epsilon_pool = [1e-7, 1e-1]
for lr_, alpha_, beta_, weight_decay_, epsilon_ in itertools.product(lr_pool, alpha_pool, beta_pool, weight_decay_pool, epsilon_pool):
    torch_result = show_pytorch(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, epsilon=epsilon_)
    mine_result = show_mine(lr=lr_, alpha=alpha_, beta=beta_, weight_decay=weight_decay_, epsilon=epsilon_)
    assert np.allclose(torch_result, mine_result)
```

### CODE REVIEW

File: onnx/defs/operator_sets.h [hunk 1]
```diff
@@ -636,6 +636,7 @@ class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, Pad);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, Gemm);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, If);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, NonMaxSuppression);
+class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, Adam);
```
Comment (chinhuang007) [inline onnx/defs/operator_sets.h]: Maybe I misunderstood, but I thought opset 11 is already part of release 1.6. Should the new operator be in opset 12 for release 1.7?

Comment (wschin) [reply onnx/defs/operator_sets.h]: Sorry.. This commit is for merging with master branch. Please only review the last commit 2d28250.

File: onnx/backend/test/data/node/test_add/model.onnx [hunk 2]
```diff
@@ -1,4 +1,4 @@
-backend-test:i
+backend-test:i
```
Comment (postrational) [inline onnx/backend/test/data/node/test_add/model.onnx]: Why are these models changed in this PR? 
`onnx/backend/test/data/node/test_add/model.onnx`
`onnx/backend/test/data/node/test_add_bcast/model.onnx`
This may be an artifact.

File: onnx/backend/test/data/node/test_add_bcast/model.onnx [hunk 3]
```diff
@@ -1,4 +1,4 @@
-backend-test:g
```
Comment (postrational) [inline onnx/backend/test/data/node/test_add_bcast/model.onnx]: Why are these models changed in this PR? 
`onnx/backend/test/data/node/test_add/model.onnx`
`onnx/backend/test/data/node/test_add_bcast/model.onnx`
This may be an artifact.

**Comment by sveta-levitan** [comment]
Wei-Sheng, please respond to Michal's comments above, plus his comments on Gitter in Operators chat room. Thank you!

**Comment by chinhuang007** [comment]
@wschin Can you please take a look and move this forward? During the TSC meeting today, the members suggest to have at least one optimizer PR merged for release 1.7. Please let us know if more time is needed.

File: onnx/backend/test/data/node/test_add_bcast/model.onnx (hunk 3 above)
Comment (wschin) [reply onnx/backend/test/data/node/test_add_bcast/model.onnx]: I will revert it. Thanks.

**Comment by wschin** [comment]
@sveta-levitan , @postrational , @chinhuang007, PR is update-to-date again. Please take a look. Thank you.

File: docs/Changelog.md [hunk 4]
```diff
@@ -14997,7 +14997,7 @@ This version of the operator has been available since version 12 of the default
   <br/>
   Where number of blocks extracted from each spatial dimension d is:
   ```
-  num_blocks[d] = floor((input_spatial_shape[d] + 2 * padding[d] − dilation[d] * (kernel_size[d] − 1) − 1) / stride[d]) + 1
+  num_blocks[d] = floor((input_spatial_shape[d] + 2 * padding[d] - dilation[d] * (kernel_size[d] - ) - 1) / stride[d]) + 1
```
Comment (chinhuang007) [inline docs/Changelog.md]: Is this change intentional?

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -114,7 +114,7 @@ Given an input of shape (N x C x D1 x D2 ... x Dn), output would be a 3-D tensor
 <br/>
 Where number of blocks extracted from each spatial dimension d is:
 ```
-num_blocks[d] = floor((input_spatial_shape[d] + 2 * padding[d] − dilation[d] * (kernel_size[d] − 1) − 1) / stride[d]) + 1
+num_blocks[d] = floor((input_spatial_shape[d] + 2 * padding[d] - dilation[d] * (kernel_size[d] - ) - 1) / stride[d]) + 1
```
Comment (chinhuang007) [inline onnx/defs/tensor/defs.cc]: This change seems to lead to other changes in docs. Maybe by accident?

Comment (wschin) [reply onnx/defs/tensor/defs.cc]: No. the `-` in the master branch is not ascii code, so my windows anaconda throws when building ONNX.

File: docs/Changelog.md (hunk 4 above)
Comment (wschin) [reply docs/Changelog.md]: Yes. I changed non-ascii `-` to ascii `-`.

File: onnx/defs/tensor/defs.cc (hunk 5 above)
Comment (chinhuang007) [reply onnx/defs/tensor/defs.cc]: Maybe I missed something. Isn't this part of documentation? And I just don't quite understand what "(kernel_size[d] - )" means. But I am certainly okay if you believe this is the right way to describe the calculation.

Comment (wschin) [reply onnx/defs/tensor/defs.cc]: I will add `1` back. Thanks!
